File: doc/src/scripts/tutorials/user-interface/UserInterface.py
import time
import krpc

conn = krpc.connect(name="User Interface Example")
canvas = conn.ui.stock_canvas

# Get the size of the game window in pixels
screen_size = canvas.rect_transform.size

# Add a panel to contain the UI elements. The user can drag it around the
# screen with the mouse.
panel = canvas.add_panel()
panel.draggable = True

# Position the panel on the left of the screen
rect = panel.rect_transform
rect.size = (200, 120)
rect.position = (110 - (screen_size[0] / 2), 0)

# Lay the contents of the panel out in a column
layout = panel.add_vertical_layout()
layout.padding = (10, 10, 10, 10)
layout.spacing = 6

# Add a button to set the throttle to maximum, with a tooltip shown while the
# mouse rests on it
button = panel.add_button("Full Throttle")
button.tooltip = "Set the throttle to maximum"
button.layout_element.preferred_size = (-1, 30)

# Add a slider to set the throttle by hand
slider = panel.add_slider()
slider.layout_element.preferred_size = (-1, 20)

# Add some text displaying the total engine thrust
text = panel.add_text("Thrust: 0 kN")
text.color = (1, 1, 1, 1)
text.size = 18

# Set up streams to monitor the button and the slider
button_clicked = conn.add_stream(getattr, button, "clicked")
slider_changed = conn.add_stream(getattr, slider, "changed")

vessel = conn.space_center.active_vessel
while True:
    # Handle the throttle button being clicked
    if button_clicked():
        vessel.control.throttle = 1
        slider.value = 1
        button.clicked = False

    # Handle the user moving the throttle slider
    if slider_changed():
        vessel.control.throttle = slider.value
        slider.changed = False

    # Update the thrust text
    text.content = "Thrust: %d kN" % (vessel.thrust / 1000)

    time.sleep(0.1)

File: doc/src/scripts/tutorials/user-interface/UserInterface.cs
using System;
using KRPC.Client;
using KRPC.Client.Services.SpaceCenter;
using KRPC.Client.Services.UI;

class UserInterface
{
    public static void Main ()
    {
        var conn = new Connection ("User Interface Example");
        var canvas = conn.UI ().StockCanvas;

        // Get the size of the game window in pixels
        var screenSize = canvas.RectTransform.Size;

        // Add a panel to contain the UI elements. The user can drag it around
        // the screen with the mouse.
        var panel = canvas.AddPanel ();
        panel.Draggable = true;

        // Position the panel on the left of the screen
        var rect = panel.RectTransform;
        rect.Size = Tuple.Create (200.0, 120.0);
        rect.Position = Tuple.Create ((110-(screenSize.Item1)/2), 0.0);

        // Lay the contents of the panel out in a column
        var layout = panel.AddVerticalLayout ();
        layout.Padding = Tuple.Create (10, 10, 10, 10);
        layout.Spacing = 6;

        // Add a button to set the throttle to maximum, with a tooltip shown
        // while the mouse rests on it
        var button = panel.AddButton ("Full Throttle");
        button.Tooltip = "Set the throttle to maximum";
        button.LayoutElement.PreferredSize = Tuple.Create (-1.0, 30.0);

        // Add a slider to set the throttle by hand
        var slider = panel.AddSlider ();
        slider.LayoutElement.PreferredSize = Tuple.Create (-1.0, 20.0);

        // Add some text displaying the total engine thrust
        var text = panel.AddText ("Thrust: 0 kN");
        text.Color = Tuple.Create (1.0, 1.0, 1.0, 1.0);
        text.Size = 18;

        // Set up streams to monitor the button and the slider
        var buttonClicked = conn.AddStream (() => button.Clicked);
        var sliderChanged = conn.AddStream (() => slider.Changed);

        var vessel = conn.SpaceCenter ().ActiveVessel;
        while (true) {
            // Handle the throttle button being clicked
            if (buttonClicked.Get ()) {
                vessel.Control.Throttle = 1;
                slider.Value = 1;
                button.Clicked = false;
            }

            // Handle the user moving the throttle slider
            if (sliderChanged.Get ()) {
                vessel.Control.Throttle = slider.Value;
                slider.Changed = false;
            }

            // Update the thrust text
            text.Content = "Thrust: " + (vessel.Thrust/1000) + " kN";

            System.Threading.Thread.Sleep (1000);
        }
    }
}
